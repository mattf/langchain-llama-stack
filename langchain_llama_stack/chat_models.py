"""Llama Stack chat models."""

import os
from typing import Any

from langchain_openai import ChatOpenAI
from pydantic import Field, SecretStr


class ChatLlamaStack(ChatOpenAI):
    """
    Llama Stack chat model integration.

    Setup:
        Install ``langchain-llama-stack`` and set optional environment variable ``LLAMA_STACK_API_KEY`` or ``LLAMA_STACK_BASE_URL``.

        .. code-block:: bash

            pip install -U langchain-llama-stack
            export LLAMA_STACK_API_KEY="your-api-key"
            export LLAMA_STACK_BASE_URL="http://my-llama-stack-disto:8321/v1/openai/v1

    Key init args — completion params:
        model: str
            Name of model to use.
        temperature: float
            Sampling temperature.
        max_tokens: Optional[int]
            Max number of tokens to generate.

    Key init args — client params:
        base_url: Optional[str]
            If not passed in will be read from env var LLAMA_STACK_BASE_URL.
        api_key: Optional[str]
            If not passed in will be read from env var LLAMA_STACK_API_KEY.
        timeout: Optional[float]
            Timeout for requests.
        max_retries: int
            Max number of retries.

    See full list of supported init args and their descriptions in the params section.

    Instantiate:
        .. code-block:: python

            from langchain_llama_stack import ChatLlamaStack

            llm = ChatLlamaStack(
                base_url="...",
                api_key="...",
                model="...",
                temperature=0,
                max_tokens=None,
                timeout=None,
                max_retries=2,
                # other params...
            )

    Invoke:
        .. code-block:: python

            messages = [
                ("system", "You are a helpful translator. Translate the user sentence to French."),
                ("human", "I love programming."),
            ]
            llm.invoke(messages)

        .. code-block:: python

            AIMessage(content='"J\'adore programmer."', additional_kwargs={}, response_metadata={'stop_reason': 'end_of_turn'}, id='run-561341ff-ac7f-41fa-bc61-13dc1c9a2ec3-0')

    Stream:
        .. code-block:: python

            for chunk in llm.stream(messages):
                print(chunk)

        .. code-block:: python

            # TODO: Example output.

        .. code-block:: python

            stream = llm.stream(messages)
            full = next(stream)
            for chunk in stream:
                full += chunk
            full

        .. code-block:: python

            AIMessageChunk(content='"J\'adore programmer."', additional_kwargs={}, response_metadata={'stop_reason': 'end_of_turn'}, id='run-3fd45daf-7488-458d-8e8f-947f7d066ef4')

    Async:
        .. code-block:: python

            await llm.ainvoke(messages)

            # stream:
            # async for chunk in (await llm.astream(messages))

            # batch:
            # await llm.abatch([messages])

        .. code-block:: python

            AIMessage(content='"J\'adore programmer."', additional_kwargs={}, response_metadata={'stop_reason': 'end_of_turn'}, id='run-fe11469c-bc41-4086-85fb-80c37a9d7efe-0')

    Tool calling:
        .. code-block:: python

            from pydantic import BaseModel, Field

            class GetWeather(BaseModel):
                '''Get the current weather in a given location'''

                location: str = Field(..., description="The city and state, e.g. Boston, MA")

            class GetPopulation(BaseModel):
                '''Get the current population in a given location'''

                location: str = Field(..., description="The city and state, e.g. Boston, MA")

            llm_with_tools = llm.bind_tools([GetWeather, GetPopulation])
            ai_msg = llm_with_tools.invoke("Which city is hotter today and which is bigger: LA or NY?")
            ai_msg.tool_calls

        .. code-block:: python

            [{'name': 'GetWeather',
              'args': {'location': 'Los Angeles, CA'},
              'id': 'chatcmpl-tool-25d24ab4522845ed8b0fe9335c5911d3',
              'type': 'tool_call'}]

        See ``ChatLlamaStack.bind_tools()`` method for more.

    Structured output:
        .. code-block:: python

            from typing import Optional

            from pydantic import BaseModel, Field

            class Joke(BaseModel):
                '''Joke to tell user.'''

                setup: str = Field(description="The setup of the joke")
                punchline: str = Field(description="The punchline to the joke")
                rating: Optional[int] = Field(description="How funny the joke is, from 1 to 10")

            structured_llm = llm.with_structured_output(Joke)
            structured_llm.invoke("Tell me a joke about cats")

        .. code-block:: python

            Joke(setup='Why did the cat join a band?', punchline='Because it wanted to be the purr-cussionist', rating=8)

        See ``ChatLlamaStack.with_structured_output()`` for more.

    JSON mode:
        .. code-block:: python

            json_llm = llm.with_structured_output(method="json_mode")
            ai_msg = json_llm.invoke("Return only a JSON object with key 'random_ints' and a value of 10 random ints in [0-99].")
            ai_msg

        .. code-block:: python

            {'random_ints': [83, 19, 91, 46, 75, 33, 28, 41, 59, 12]}

    Image input:
        .. code-block:: python

            import base64
            import httpx
            from langchain_core.messages import HumanMessage

            image_url = "https://upload.wikimedia.org/wikipedia/commons/thumb/d/dd/Gfp-wisconsin-madison-the-nature-boardwalk.jpg/2560px-Gfp-wisconsin-madison-the-nature-boardwalk.jpg"
            image_data = base64.b64encode(httpx.get(image_url).content).decode("utf-8")
            message = HumanMessage(
                content=[
                    {"type": "text", "text": "describe the weather in this image"},
                    {
                        "type": "image_url",
                        "image_url": {"url": f"data:image/jpeg;base64,{image_data}"},
                    },
                ],
            )
            ai_msg = llm.invoke([message])
            ai_msg.content

        .. code-block:: python

            'The weather in this image is sunny and warm, with a clear blue sky and fluffy white clouds. The sun is shining down on the boardwalk and the grass, casting a warm glow over the entire scene. The air is filled with a gentle breeze that rustles the leaves of the trees and carries the sweet scent of blooming flowers. Overall, the atmosphere is peaceful and serene, inviting the viewer to step into the idyllic world depicted in the image.'

    Token usage:
        .. code-block:: python

            ai_msg = llm.invoke(messages)
            ai_msg.usage_metadata

        .. code-block:: python

            {'input_tokens': 32, 'output_tokens': 17, 'total_tokens': 49}

    Logprobs:
        .. code-block:: python

            logprobs_llm = llm.bind(logprobs=True)  # or logprobs=3 for top 3
            ai_msg = logprobs_llm.invoke(messages)
            ai_msg.response_metadata["logprobs"]

        .. code-block:: python

            [{'"': -2.4698164463043213},
             {'J': -0.31870245933532715},
             {"'": -0.15102243423461914},
             {'ad': -0.004451931454241276},
             {'ore': -0.0009182137437164783},
             {' programmer': -1.1367093324661255},
             {'."': -0.14114761352539062}]

    Response metadata
        .. code-block:: python

            ai_msg = llm.invoke(messages)
            ai_msg.response_metadata

        .. code-block:: python

            {'stop_reason': 'end_of_turn'}

    """  # noqa: E501

    openai_api_base: str | None = Field(
        alias="base_url",
        default_factory=lambda: os.environ.get(
            "LLAMA_STACK_BASE_URL", "http://localhost:8321/v1/openai/v1"
        ),
    )
    "Loaded from LLAMA_STACK_BASE_URL environment variable if not provided."
    openai_api_key: SecretStr | None = Field(
        alias="api_key",
        default_factory=lambda: SecretStr(
            os.environ.get("LLAMA_STACK_API_KEY", "NO_API_KEY")
        ),
    )
    "Loaded from LLAMA_STACK_API_KEY environment variable if not provided."

    @property
    def _llm_type(self) -> str:
        """Return type of chat model."""
        return "llama-stack-chat"

    @property
    def _identifying_params(self) -> dict[str, Any]:
        return {
            "model_name": self.model_name,
        }

    @classmethod
    def is_lc_serializable(cls) -> bool:
        """Return whether this model can be serialized by LangChain."""
        return False
