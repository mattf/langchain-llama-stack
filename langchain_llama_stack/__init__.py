from importlib import metadata

from langchain_llama_stack.chat_models import ChatLlamaStack

# Import safety hooks
from langchain_llama_stack.input_output_safety_moderation_hooks import (
    SafeLLMWrapper,
    create_safe_llm,
    create_safety_hook,
)
from langchain_llama_stack.safety import LlamaStackSafety, SafetyResult

# from langchain_llama_stack.document_loaders import LlamaStackLoader
# from langchain_llama_stack.embeddings import LlamaStackEmbeddings
# from langchain_llama_stack.retrievers import LlamaStackRetriever
# from langchain_llama_stack.toolkits import LlamaStackToolkit
# from langchain_llama_stack.tools import LlamaStackTool
# from langchain_llama_stack.vectorstores import LlamaStackVectorStore

try:
    __version__ = metadata.version(__package__)
except metadata.PackageNotFoundError:
    # Case where package metadata is not available.
    __version__ = ""
del metadata  # optional, avoids polluting the results of dir(__package__)

__all__ = [
    "ChatLlamaStack",
    "LlamaStackSafety",
    "SafetyResult",
    "SafeLLMWrapper",
    "create_safe_llm",
    "create_safety_hook",
    "__version__",
    # Future components (commented out until implemented):
    # "LlamaStackVectorStore",
    # "LlamaStackEmbeddings",
    # "LlamaStackTelemetry",
    # "LlamaStackLoader",
]
