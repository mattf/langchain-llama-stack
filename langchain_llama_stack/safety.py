"""LlamaStack Safety integration for LangChain."""

import logging
import os
from typing import Any, Optional

from pydantic import BaseModel

# Set up logging
logger = logging.getLogger(__name__)
logging.basicConfig(level=logging.INFO)

try:
    from llama_stack_client import (  # type: ignore
        AsyncLlamaStackClient,
        LlamaStackClient,
    )
except ImportError:
    AsyncLlamaStackClient = None  # type: ignore
    LlamaStackClient = None  # type: ignore


class SafetyResult(BaseModel):
    """Result from safety check."""

    is_safe: bool
    violations: list[dict[str, Any]] = []
    confidence_score: Optional[float] = None
    explanation: Optional[str] = None
    metadata: Optional[dict[str, Any]] = None


class LlamaStackSafety:
    """
    Llama Stack safety and moderation integration.

    Key init args — safety params:
        model: str
            Name of safety model to use for safety and moderation\
             (default: "llama-guard").

    Key init args — client params:
        base_url: str
            Llama Stack server URL (default: "http://localhost:8321").
        timeout: Optional[float]
            Timeout for requests (default: 30.0).
        max_retries: int
            Max number of retries (default: 2).

    Instantiate:
        .. code-block:: python

            from langchain_llama_stack import LlamaStackSafety

            # For local Llama Stack (no API key needed)
            safety = LlamaStackSafety(
                base_url="http://localhost:8321", shield_type="llama_guard"
            )

            # For remote Llama Stack (with API key)
            safety = LlamaStackSafety(
                base_url="http://remote-llama-stack:8321",
                shield_type="llama_guard",
            )

    Check content safety:
        .. code-block:: python

            result = safety.check_content_safety("This is some text to check")
            print(result.is_safe)
            print(result.violations)

        .. code-block:: python

            SafetyResult(is_safe=True, violations=[], confidence_score=0.95)

    Async:
        .. code-block:: python

            result = await safety.acheck_content_safety("Text to check")

        .. code-block:: python

            SafetyResult(is_safe=True, violations=[])
    """

    def __init__(
        self,
        base_url: Optional[str] = None,
        model: str = "ollama/llama-guard3:8b",
        timeout: Optional[float] = 30.0,
        max_retries: int = 2,
    ):
        """Initialize LlamaStackSafety."""
        self.base_url = base_url or os.environ.get(
            "LLAMA_STACK_BASE_URL", "http://localhost:8321"
        )
        self.model = model
        self.timeout = timeout
        self.max_retries = max_retries

        # Clients will be initialized lazily when needed
        self.client: Optional[LlamaStackClient] = None
        self.async_client: Optional[AsyncLlamaStackClient] = None

    def _get_client_kwargs(self) -> dict[str, Any]:
        """Get common client kwargs."""
        client_kwargs = {
            "base_url": self.base_url,
            "timeout": self.timeout,
            "max_retries": self.max_retries,
        }
        return client_kwargs

    def _initialize_client(self) -> None:
        """Initialize the Llama Stack client."""
        if self.client is None and LlamaStackClient is not None:
            self.client = LlamaStackClient(**self._get_client_kwargs())

    def _initialize_async_client(self) -> None:
        """Initialize the async Llama Stack client."""
        if self.async_client is None and AsyncLlamaStackClient is not None:
            self.async_client = AsyncLlamaStackClient(**self._get_client_kwargs())

    def list_shields(self) -> list[str]:
        """List all available shields from LlamaStack server."""
        try:
            if self.client is None:
                self._initialize_client()

            if self.client is None:
                return []

            # Get shields
            shields_response = self.client.shields.list()
            # Handle both list and object with .data attribute
            if hasattr(shields_response, "data"):
                shields_data = shields_response.data
            else:
                shields_data = shields_response

            # Return just the identifiers for easy checking
            shield_ids = [shield.identifier for shield in shields_data]
            return shield_ids

        except Exception as e:
            logger.error(f"Error listing shields: {e}")
            return []

    def check_content_safety(
        self, content: str, content_type: str = "text", **kwargs: Any
    ) -> SafetyResult:
        """
        Check content safety using Llama Stack shields.

        Args:
            content: The content to check for safety
            content_type: Type of content (default: "text")
            **kwargs: Additional parameters for safety checking

        Returns:
            SafetyResult with safety assessment
        """
        # logger.info(f"Starting safety check for content: '{content[:50]}...'")
        # logger.info(f"Using shield_type: {self.shield_type}")
        # logger.info(f"Base URL: {self.base_url}")

        # Check if LlamaStackClient is available
        if LlamaStackClient is None:
            logger.error("LlamaStackClient is None - client not available")
            return SafetyResult(
                is_safe=True,
                violations=[],
                explanation=(
                    "LlamaStackClient not available - install llama-stack-client"
                ),
            )

        logger.info("LlamaStackClient is available")

        if self.client is None:
            logger.info("Client is None, initializing...")
            self._initialize_client()

        try:
            # Ensure client is not None after initialization
            if self.client is None:
                logger.error("Client is still None after initialization")
                return SafetyResult(
                    is_safe=True,
                    violations=[],
                    explanation="LlamaStack client not initialized",
                )

            logger.info("Client initialized successfully")
            logger.info(f"Making API call with model: {self.model}")

            # Use the moderations.create method
            response = self.client.moderations.create(
                input=content,
                model=self.model,
                **kwargs,
            )

            # Parse OpenAI-compatible moderation response
            is_safe = True
            violations = []
            confidence_score = None
            explanation = None
            metadata = {}

            # Extract top-level metadata
            if hasattr(response, "id"):
                metadata["id"] = response.id
            if hasattr(response, "model"):
                metadata["model"] = response.model

            # Process results array
            if hasattr(response, "results") and response.results:
                # Use the first result (typically there's only one)
                result = response.results[0]

                # Check if content is flagged
                is_flagged = getattr(result, "flagged", False)
                is_safe = not is_flagged

                # Extract user message if available
                if hasattr(result, "user_message"):
                    explanation = result.user_message

                # Extract metadata
                if hasattr(result, "metadata"):
                    result_metadata = result.metadata
                    if isinstance(result_metadata, dict):
                        # Ensure all values are strings for type safety
                        str_metadata = {k: str(v) for k, v in result_metadata.items()}
                        metadata.update(str_metadata)
                    else:
                        # Handle object-style metadata
                        for attr in dir(result_metadata):
                            if not attr.startswith("_"):
                                value = getattr(result_metadata, attr)
                                if not callable(value):
                                    metadata[attr] = str(value)

                # Process flagged categories
                if is_flagged and hasattr(result, "categories"):
                    categories = result.categories
                    category_scores = getattr(result, "category_scores", {})

                    if isinstance(categories, dict):
                        # Dictionary format: {"violence": True, "hate": False, ...}
                        flagged_categories = [
                            cat for cat, flagged in categories.items() if flagged
                        ]
                    else:
                        # Object format: result.categories.violence = True
                        flagged_categories = [
                            attr
                            for attr in dir(categories)
                            if not attr.startswith("_")
                            and getattr(categories, attr, False)
                        ]

                    # Create violations from flagged categories
                    for category in flagged_categories:
                        score = None
                        if isinstance(category_scores, dict):
                            score = category_scores.get(category)
                        else:
                            score = getattr(category_scores, category, None)

                        violations.append(
                            {"category": category, "score": score, "flagged": True}
                        )

                # Extract confidence score if available
                # (some implementations may have this)
                if hasattr(result, "confidence_score"):
                    confidence_score = result.confidence_score
            else:
                # Fallback: if no results, treat as safe
                logger.warning("No results found in moderation response")

            # logger.info(f"Final result - is_safe:
            # {is_safe},
            # violations: {violations}")

            return SafetyResult(
                is_safe=is_safe,
                violations=violations,
                confidence_score=confidence_score,
                explanation=explanation,
            )
        except Exception as e:
            logger.error(f"Exception occurred during safety check: {str(e)}")
            logger.error(f"Exception type: {type(e)}")
            # Return safe by default on error, but log the issue
            return SafetyResult(
                is_safe=True,
                violations=[],
                explanation=f"Safety check failed: {str(e)}",
            )

    async def acheck_content_safety(
        self, content: str, content_type: str = "text", **kwargs: Any
    ) -> SafetyResult:
        """
        Async check content safety using Llama Stack safety shields.

        Args:
            content: The content to check for safety
            content_type: Type of content (text, image, etc.)
            **kwargs: Additional parameters for safety check

        Returns:
            SafetyResult with safety assessment
        """
        if not self.async_client:
            self._initialize_async_client()

        try:
            # Ensure async client is not None
            if self.async_client is None:
                raise ValueError("LlamaStack async client not initialized")

            # Use the AsyncLlamaStackClient.moderations.create method
            response = await self.async_client.moderations.create(
                input=content,
                model=self.model,
                **kwargs,
            )

            # Parse OpenAI-compatible moderation response (same logic as sync version)
            is_safe = True
            violations = []
            confidence_score = None
            explanation = None
            metadata = {}

            # Extract top-level metadata
            if hasattr(response, "id"):
                metadata["id"] = response.id
            if hasattr(response, "model"):
                metadata["model"] = response.model

            # Process results array
            if hasattr(response, "results") and response.results:
                # Use the first result (typically there's only one)
                result = response.results[0]

                # Check if content is flagged
                is_flagged = getattr(result, "flagged", False)
                is_safe = not is_flagged

                # Extract user message if available
                if hasattr(result, "user_message"):
                    explanation = result.user_message

                # Extract metadata
                if hasattr(result, "metadata"):
                    result_metadata = result.metadata
                    if isinstance(result_metadata, dict):
                        # Ensure all values are strings for type safety
                        str_metadata = {k: str(v) for k, v in result_metadata.items()}
                        metadata.update(str_metadata)
                    else:
                        # Handle object-style metadata
                        for attr in dir(result_metadata):
                            if not attr.startswith("_"):
                                value = getattr(result_metadata, attr)
                                if not callable(value):
                                    metadata[attr] = str(value)

                # Process flagged categories
                if is_flagged and hasattr(result, "categories"):
                    categories = result.categories
                    category_scores = getattr(result, "category_scores", {})

                    if isinstance(categories, dict):
                        # Dictionary format: {"violence": True, "hate": False, ...}
                        flagged_categories = [
                            cat for cat, flagged in categories.items() if flagged
                        ]
                    else:
                        # Object format: result.categories.violence = True
                        flagged_categories = [
                            attr
                            for attr in dir(categories)
                            if not attr.startswith("_")
                            and getattr(categories, attr, False)
                        ]

                    # Create violations from flagged categories
                    for category in flagged_categories:
                        score = None
                        if isinstance(category_scores, dict):
                            score = category_scores.get(category)
                        else:
                            score = getattr(category_scores, category, None)

                        violations.append(
                            {"category": category, "score": score, "flagged": True}
                        )

                # Extract confidence score if available
                # (some implementations may have this)
                if hasattr(result, "confidence_score"):
                    confidence_score = result.confidence_score
            else:
                # Fallback: if no results, treat as safe
                logger.warning("No results found in async moderation response")

            return SafetyResult(
                is_safe=is_safe,
                violations=violations,
                confidence_score=confidence_score,
                explanation=explanation,
            )
        except Exception as e:
            # Return safe by default on error, but log the issue
            return SafetyResult(
                is_safe=True,
                violations=[],
                explanation=f"Async safety check failed: {str(e)}",
            )
