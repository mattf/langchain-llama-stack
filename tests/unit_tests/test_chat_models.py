from typing import Tuple, Type

from langchain_tests.unit_tests import ChatModelUnitTests

from langchain_llama_stack.chat_models import ChatLlamaStack


class TestChatLlamaStackUnit(ChatModelUnitTests):
    @property
    def chat_model_class(self) -> Type[ChatLlamaStack]:
        return ChatLlamaStack

    @property
    def chat_model_params(self) -> dict:
        return {
            "model": "not-a-model",
        }

    @property
    def init_from_env_params(self) -> Tuple[dict, dict, dict]:
        """(tuple) environment variables, additional initialization args, and expected
        instance attributes for testing initialization from environment variables."""
        return (
            {
                "LLAMA_STACK_API_KEY": "test-key",  # we need this for test_serdes
            },
            {
                "model": "test-a-model",
            },
            {
                # when using LLAMA_STACK_API_KEY env, the key is not stored on
                # ChatLlamaStack, it is stored in the underlying LlamaStackClient.
                # therefore, we expect the key to be None on ChatLlamaStack and
                # cannot reasonably check it.
                # "api_key": None,
            },
        )

    # TODO(mf): re-enable when we can extend the test suite
    # import pytest_mock
    # import pytest
    # from llama_stack_client import NOT_GIVEN
    # from langchain_core.language_models import BaseChatModel
    # from langchain_core.tools import BaseTool
    # from llama_stack_client.resources.inference import InferenceResource
    # from llama_stack_client.types import (
    #     ChatCompletionResponse,
    #     CompletionMessage,
    # )
    # @pytest.mark.parametrize(
    #     "tool_choice, expected_tool_config",
    #     [
    #         ("auto", {"tool_choice": "auto"}),
    #         ("required", {"tool_choice": "required"}),
    #         ("none", {"tool_choice": "none"}),
    #         ("xyz", {"tool_choice": "xyz"}),
    #         ("any", {"tool_choice": "auto"}),
    #         (None, NOT_GIVEN),
    #     ],
    #     ids=[
    #         "auto",
    #         "required",
    #         "none",
    #         "xyz",
    #         "any",
    #         "None",
    #     ],
    # )
    # def test_bind_tools_tool_choice(
    #     self,
    #     model: BaseChatModel,
    #     mocker: pytest_mock.MockFixture,
    #     my_adder_tool: BaseTool,
    #     tool_choice: str,
    #     expected_tool_config: str,
    # ) -> None:
    #     mock_chat_completion = mocker.patch.object(
    #         InferenceResource,
    #         "chat_completion",
    #         return_value=ChatCompletionResponse(
    #             completion_message=CompletionMessage(
    #                 role="assistant",
    #                 content="Hello, world!",
    #                 stop_reason="end_of_turn",
    #             ),
    #         ),
    #     )

    #     bound_model = model.bind_tools(tools=[my_adder_tool], tool_choice=tool_choice)
    #     bound_model.invoke("Hello, world!")

    #     mock_chat_completion.assert_called_once()
    #     _, kwargs = mock_chat_completion.call_args
    #     assert "tool_config" in kwargs
    #     assert kwargs["tool_config"] == expected_tool_config
