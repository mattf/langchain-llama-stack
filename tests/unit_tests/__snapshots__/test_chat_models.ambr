# serializer version: 1
# name: TestChatLlamaStackUnit.test_serdes[serialized]
  dict({
    'id': list([
      'langchain_llama_stack',
      'chat_models',
      'ChatLlamaStack',
    ]),
    'kwargs': dict({
      'api_key': dict({
        'id': list([
          'LLAMA_STACK_API_KEY',
        ]),
        'lc': 1,
        'type': 'secret',
      }),
      'max_retries': 2,
      'max_tokens': 100,
      'model': 'not-a-model',
      'model_name': 'not-a-model',
      'temperature': 0.0,
      'timeout': 60,
    }),
    'lc': 1,
    'name': 'ChatLlamaStack',
    'type': 'constructor',
  })
# ---
